library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```
@@ -0,0 +1,304 @@
+          /* If the last parameter is a default parameter, the method can also
+           * be implemented by two methods, one with and one without the
+           * parameter. Note that this does not generalize to more parameters
+           * due to named parameters.
+           */
+          val pinit = params.init
+          val noLastDefault = pinit :+ params.last.copy(isDefault = false)
+
+          find(jsMem.copy(params = pinit)) &&
+          find(jsMem.copy(params = noLastDefault))
+
+        case _ => false
+      }
+
+      find(jsMem) || hasConformingOverloads
+    }
+
+    /** Members that a facade type defines */
+    private def rawJSMembers(tpe: Type): Iterable[(String, JSMember)] = {
+
+      def isValidMember(mem: Symbol) = (
+          !JSObjectAncestors(mem.owner) &&
+          !mem.isConstructor &&
+          mem.isMethod &&
+          !mem.asTerm.isParamWithDefault
+      )
+
+      for (mem <- tpe.members if isValidMember(mem)) yield {
+        val (name, explicit) = jsName(mem.asMethod)
+
+        // apply is a special case, since when it is not explicitly specified,
+        // it will result in an actual function application in JavaScript
+        // Scala defined types cannot support this --> fail
+        if (name == "apply" && !explicit) {
+          c.abort(c.enclosingPosition, s"$tpe defines an apply method. This " +
+            "cannot be implemented by any Scala exported type, since it " +
+            "would need to chain Function's prototype.")
+        }
+
+        (name, jsMemberFor(tpe, mem.asMethod))
+      }
+    }
+
+    /** Returns the JavaScript name of a method and whether it is specified
+     *  explicitly or not.
+     */
+    private def jsName(sym: MethodSymbol): (String, Boolean) = {
+      val annot = memberAnnotations(sym).find(annotIs(_, JSNameAnnotation))
+      val name = annot.flatMap(annotStringArg).getOrElse(defaultName(sym))
+      (name, annot.isDefined)
```

I think this is a bad API, even for a private method. The explicitness should be irrelevant when returning from this method. The `apply` case (and the `JSBracketSelect` and `JSBracketCall` cases) should either fail early in here and not leak, or return a value of the following ADT (or something similar):
```scala
sealed abstract class JSMemberSelection
case class JSMemberSelectByName(name: String) extends JSMemberSelection
case object JSMemberCall extends JSMemberSelection
case object JSMemberBracketSelect extends JSMemberSelection
case object JSMemberBracketCall extends JSMemberSelection
```
or its useful contraction as `Option[String]` ^^ where `None` is used for the three last cases.